### PR TITLE
support multiple bind records (fie nodelocaldns  test regression)

### LINKED ISSUE #95573: *-dns-nodecache jobs are failing
**Which jobs are failing**:
gci-gce-serial-kube-dns-nodecache

**Which test(s) are failing**:
N/A

**Since when has it been failing**:
10/13

**Testgrid link**:
https://k8s-testgrid.appspot.com/sig-network-gce#gci-gce-serial-kube-dns-nodecache&width=5

**Reason for failure**:
Cluster failed to start:
```
I1014 22:01:41.572]   Oct 14 22:01:41.512: Error waiting for all pods to be running and ready: 4 / 24 pods in namespace "kube-system" are NOT in RUNNING and READY state in 10m0s
```
**Anything else we need to know**:
/cc @aojea @jayunit100 @prameshj

### PR DESCRIPTION
**Pull request by jayunit100**
Fixes #95573

### CODE REVIEW

**Comment by jayunit100** [comment]
Thanks to  @prameshj for repro'ing on GCE and  @aojea  for digging out the diff between the DNS records in the working job  ! Heres the fix.

So after readint through it all, it appears that, 

- When we removed the `__PILLAR__` stuff originally, we didnt take into account the fact that nodelocaldns actually needed to write 2 records for the `bind` lines .  In THAT case, it turns out the two records were both getting replaced by the single, FIRST IP address .  The result is that the FIRST ip address to do this substitution (the server)  winds up being the first thing in the `bind` directive. 
- I think theres some magic where the node local dns winds up creating or adding an IP somehow for the local_dns value, and then it needs to listen on this IP.  If that IP isnt create when the container starts, an infinite reboot happens, which eventually causes the job to faill 

The solution is of course to just ... NOT USE A REGULAR EXPRESSION :).

**Comment by jayunit100** [comment]
testing this now in GKE, if it passes ill ping folks to merge if possible so we can get these green again by EOD possibly

File: hack/local-up-cluster.sh [hunk 1]
```diff
@@ -905,9 +905,14 @@ function start_kubedns {
 function start_nodelocaldns {
   cp "${KUBE_ROOT}/cluster/addons/dns/nodelocaldns/nodelocaldns.yaml" nodelocaldns.yaml
   # .* because of the __PILLLAR__ references that eventually will be removed
-  ${SED} -i -e "s/_.*_DNS__DOMAIN__/${DNS_DOMAIN}/g" nodelocaldns.yaml
-  ${SED} -i -e "s/_.*_DNS__SERVER__/${DNS_SERVER_IP}/g" nodelocaldns.yaml
-  ${SED} -i -e "s/_.*_LOCAL__DNS__/${LOCAL_DNS_IP}/g" nodelocaldns.yaml
+  ${SED} -i -e "s/__DNS__DOMAIN__/${DNS_DOMAIN}/g" nodelocaldns.yaml
```
Comment (prameshj) [inline hack/local-up-cluster.sh:908]: Won't this partially replace the string - "__PILLAR__DNS__DOMAIN__"? Can we remove the non-PILLAR lines unless we actually need it?

Comment (jayunit100) [reply hack/local-up-cluster.sh:908]: +1 yeah good point haven't tested yet

**Comment by jayunit100** [comment]
Ok, here's the new core file.   I finally got my GCE reds setup so I could test it properly... 

In response to your comment above, your right, once I finally ran it it was obvious :), need to make sure and 

1) first substitute the __PILLAR__A_B_C  expressions
2) then substitute A_B_C expressions

This results in both IP addresses properly flowing to coredns.

```
2020/10/16 01:06:02 [INFO] Using config file:                                                                                                                                                                                                    
cluster.local:53 {                                                                                                                                                                                                                               
    errors                                                                                                                                                                                                                                       
    cache {                                                                                                                                                                                                                                      
            success 9984 30                                                                                                                                                                                                                      
            denial 9984 5                                                                                                                                                                                                                        
    }                                                                                                                                                                                                                                            
    reload                                                                                                                                                                                                                                           loop                                                                                                                
    bind 169.254.20.10 10.0.0.10                                                                                                                                                                                                                     forward . 10.0.178.227 {                                                                                            
            force_tcp                                                                                                                                                                                                                                }                                                                                                                   
    prometheus :9253                                                                                                                                                                                                                                 health 169.254.20.10:8080                                                                                           
    }                                                                                                                                                                                                                                            in-addr.arpa:53 {                                                                                                       
    errors                                                                                                                                                                                                                                           cache 30                                                                                                            
    reload                                                                      
    loop                                                                        
    bind 169.254.20.10 10.0.0.10                                                
    forward . 10.0.178.227 {                                                    
            force_tcp                                                           
    }                                                                           
    prometheus :9253                                                            
    }                                                                           
ip6.arpa:53 {                                                                   
    errors                                                                      
    cache 30                                                                    
    reload                                                                      
    loop                                                                        
    bind 169.254.20.10 10.0.0.10                                                
    forward . 10.0.178.227 {                                                    
            force_tcp                                                           
    }                                                                           
    prometheus :9253                                                            
    }                                                                           
.:53 {                                                                          
    errors                                                                      
    cache 30                                                                    
    reload                                                                      
    loop                                                                        
    bind 169.254.20.10 10.0.0.10                                                
    forward . /etc/resolv.conf                                                  
    prometheus :9253                                                            
    }                                                                           
2020/10/16 01:06:02 [INFO] Updated Corefile with 0 custom stubdomains and upstream servers /etc/resolv.conf
2020/10/16 01:06:02 [INFO] Using config file:                                   
cluster.local:53 {                                                              
    errors                                                                      
    cache {                                                                     
            success 9984 30                                                     
            denial 9984 5                                                       
    }                                                                           
    reload                                                                      
    loop                                                                        
    bind 169.254.20.10 10.0.0.10                                                
    forward . 10.0.178.227 {                                                    
            force_tcp                                                           
    }                                                                           
    prometheus :9253                                                            
    health 169.254.20.10:8080                                                   
    }                                                                           
in-addr.arpa:53 {                                                               
    errors                                                                      
    cache 30                                                                    
    reload                                                                      
    loop                                                                        
    bind 169.254.20.10 10.0.0.10                                                
    forward . 10.0.178.227 {                                                    
            force_tcp                                                           
    }                                                                           
    prometheus :9253                                                             
```

File: hack/local-up-cluster.sh (hunk 1 above)
Comment (jayunit100) [reply hack/local-up-cluster.sh]: you were right on this, I made sure its ordered properly

**Comment by prameshj** [comment]
> Ok, here's the new core file. I finally got my GCE reds setup so I could test it properly...
> 
> In response to your comment above, your right, once I finally ran it it was obvious :), need to make sure and
> 
> 1. first substitute the __PILLAR__A_B_C  expressions
> 2. then substitute A_B_C expressions

Thanks for checking. Do we need this second substitution? As I understand right, they will only be needed when nodelocaldns.yaml replaces all the "__PILLAR__" variables.
If there isn't a strong reason to keep it, my preference is to add a comment and remove those extra sed commands. 
> 
> This results in both IP addresses properly flowing to coredns....
> 
> **So I think its haPPY NOW :)**
> 
> shall we merge and monitor over the next day ?
> 
> ```
> 2020/10/16 01:06:02 [INFO] Using config file:                                                                                                                                                                                                    
> cluster.local:53 {                                                                                                                                                                                                                               
>     errors                                                                                                                                                                                                                                       
>     cache {                                                                                                                                                                                                                                      
>             success 9984 30                                                                                                                                                                                                                      
>             denial 9984 5                                                                                                                                                                                                                        
>     }                                                                                                                                                                                                                                            
>     reload                                                                                                                                                                                                                                           loop                                                                                                                
>     bind 169.254.20.10 10.0.0.10                                                                                                                                                                                                                     forward . 10.0.178.227 {                                                                                            
>             force_tcp                                                                                                                                                                                                                                }                                                                                                                   
>     prometheus :9253                                                                                                                                                                                                                                 health 169.254.20.10:8080                                                                                           
>     }                                                                                                                                                                                                                                            in-addr.arpa:53 {                                                                                                       
>     errors                                                                                                                                                                                                                                           cache 30                                                                                                            
>     reload                                                                      
>     loop                                                                        
>     bind 169.254.20.10 10.0.0.10                                                
>     forward . 10.0.178.227 {                                                    
>             force_tcp                                                           
>     }                                                                           
>     prometheus :9253                                                            
>     }                                                                           
> ip6.arpa:53 {                                                                   
>     errors                                                                      
>     cache 30                                                                    
>     reload                                                                      
>     loop                                                                        
>     bind 169.254.20.10 10.0.0.10                                                
>     forward . 10.0.178.227 {                                                    
>             force_tcp                                                           
>     }                                                                           
>     prometheus :9253                                                            
>     }                                                                           
> .:53 {                                                                          
>     errors                                                                      
>     cache 30                                                                    
>     reload                                                                      
>     loop                                                                        
>     bind 169.254.20.10 10.0.0.10                                                
>     forward . /etc/resolv.conf                                                  
>     prometheus :9253                                                            
>     }                                                                           
> 2020/10/16 01:06:02 [INFO] Updated Corefile with 0 custom stubdomains and upstream servers /etc/resolv.conf
> 2020/10/16 01:06:02 [INFO] Using config file:                                   
> cluster.local:53 {                                                              
>     errors                                                                      
>     cache {                                                                     
>             success 9984 30                                                     
>             denial 9984 5                                                       
>     }                                                                           
>     reload                                                                      
>     loop                                                                        
>     bind 169.254.20.10 10.0.0.10                                                
>     forward . 10.0.178.227 {                                                    
>             force_tcp                                                           
>     }                                                                           
>     prometheus :9253                                                            
>     health 169.254.20.10:8080                                                   
>     }                                                                           
> in-addr.arpa:53 {                                                               
>     errors                                                                      
>     cache 30                                                                    
>     reload                                                                      
>     loop                                                                        
>     bind 169.254.20.10 10.0.0.10                                                
>     forward . 10.0.178.227 {                                                    
>             force_tcp                                                           
>     }                                                                           
>     prometheus :9253                                                             
> ```

**Comment by jayunit100** [comment]
My thinking was... in the future if we add or update this file somehow, it will support the same substitution matrix  as other dns files.

I don't feel strongly there... if you still think it's a bad idea I'll remove the extra substitution steps.

Let me know, thanks @prameshj

**Comment by jayunit100** [comment]
@prameshj removed the extra substitutions to expedite merging and getting the node local jobs green again :)

**Comment by aojea** [comment]
ping @dims 

> dims # for local-up-cluster related files

**Pull request merged**